src/mlpack/tests/gmm_test.cpp

```diff
@@ -894,4 +895,45 @@ BOOST_AUTO_TEST_CASE(DiagonalGMMTrainTest)
   }
 }
 
+/**
+ * Test training a model on only one Gaussian (randomly generated) in two
+ * dimensions.  We will vary the dataset size from small to large.  The EM
+ * algorithm is used for training the DiagonalGMM.
+ */
+BOOST_AUTO_TEST_CASE(DiagonalGMMTrainEMOneGaussian)
+{
+  for (size_t iterations = 0; iterations < 4; iterations++)
+  {
+    // Determine random covariance and mean.
+    arma::vec mean;
+    mean.randu(2);
+    arma::vec covar;
+    covar.randu(2);
+
+    arma::mat data;
+    data.randn(2 /* dimension */, 150 * pow(10, (iterations / 3.0)));
+
+    // Now apply mean and covariance.
+    data.row(0) *= covar(0);
+    data.row(1) *= covar(1);
+
+    data.row(0) += mean(0);
+    data.row(1) += mean(1);
+
+    // Now, train the model.
+    DiagonalGMM gmm(1, 2);
+    gmm.Train(data, 10);
+
+    arma::vec actualMean = arma::mean(data, 1);
+    arma::mat actualCovar = arma::diagmat(
+        arma::ccov(data, 1 /* biased estimator */));
+
+    // Check the model to see that it is correct.
+    CheckMatrices(gmm.Component(0).Mean(), actualMean);
+    CheckMatrices(gmm.Component(0).Covariance(), actualCovar);
+
+    BOOST_REQUIRE_CLOSE(gmm.Weights()[0], 1.0, 1e-5);
```

Comment:
I've had a lot of problems in the past with the GMM and HMM tests because they aren't robust to different random seeds (since the data is generated randomly).  You can test against different random seeds by adding `math::RandomSeed(std::time(NULL));` at the top of the test, then compiling and running from a shell a loop like this:

```
$ i=0; while(true); do echo $i; bin/mlpack_test -t GMMTest/DiagonalGMMTrainEMOneGaussian 2>&1 | grep 'fatal error'; sleep 1; i=$(($i + 1)); done
```

I'll often try to make sure that runs for something like 1000 iterations without any failures.  (That can take a little while, so I will start it over lunch in a screen or tmux session or something so I can look at the output when I come back. :))

Reply:
It was really really helpful to me, thank you so much. :)